### ISSUE TITLE
Consider letting harvesters search from refinery and not last harvested position

### ISSUE THREAD

**Issue by abcdefg30** [OpenRA/OpenRA#17046]
Mainly when it was not an explicit order, and we should at least do it for the AI imho.

I admit I'm not up to speed with the current AI discussion and this was probably already suggested somewhen. I don't think this is required for the next release, but should be considered for at least next+1, as this can lead to bogus looking behaviour as illustrated in the following case:
![grafik](https://user-images.githubusercontent.com/7704140/64345984-98416000-cff1-11e9-98e3-7ae7eb503594.png)
The gems in the top left ran empty while the harvester was not yet full, making it move to the ore on the bottom right. After a while the gems grew back, but the harvester still long-distance mined the ore (without ever having received an order).
![grafik](https://user-images.githubusercontent.com/7704140/64346057-bd35d300-cff1-11e9-8a22-f4076c2f609d.png)

I think the behaviour would improve in general if we didn't harvest from the last position but refinery location unless an order was given.

**Comment by matjaeck** [comment]
> The gems in the top left ran empty while the harvester was not yet full, making it move to the ore on the bottom right.

I have suggested a while ago that a delay before the harvester scans for new ore might help (the harvester would idle during that time), so that ore of a oremine can regrow.

**Comment by GraionDilach** [comment]
The proposal would suck more to the AI that what it does currently.

In the example screenshot, a new harvester built from the factory with a lucky rallypoint would hop to the gems and if the constant reprioritizing would take place, then both miners would limit themselves to the gemmine even after it gets exhausted and would lead to single-bail runs in the long term.

Distance mining sucks, sure, but it is less punishing for the AI economy than mining single patches, something it did in the previous release constantly. If the AI employs more harvesters, then these harvesters are able to finish a patch and not stop there but move on to another, giving time the patch to properly regrow.

A true solution would be that the AI actively tries to distribute harvesters between patches but that is a lot more effort than what the options are currently. If long distance mining is a real issue, then it's better to increase the amount of harvesters the AI operates with on the short term.

**Comment by netnazgul** [comment]
How do you differentiate between an explicit order to gather long distance rather than something close to the ref from a ref scan result? I assume a player expects harvesters to go to the pointed location instead of them trying to be smart.

Another question is - which ref to use as a scan source in this case? Slightly out of scope for the issue, but the "3 harvesters per refinery" implicit rule exhibits too much weird behaviour nowadays, i.e. 4 harvesters working near a refinery results in one of them switching to any other refinery on the map to unload even if that is across the whole map distance. Getting back to the original question - there might be a refinery near to that ore patch that a player expects ore truck to unload to after it mined the ore. Which refinery will it actually pick for unload? Which refinery is used for a resource scan?

**Comment by abcdefg30** [comment]
@GraionDilach I don't think you understand what is proposed here. The harvesters would still transfer if they ran out of gems to mine from the patch. Otherwise this would be outright broken if no mine is present at the initial patch.

> giving time the patch to properly regrow.

No, that is exactly what you *don't* want to happen. Having gems sit unmined directly next to your refinery is a huge waste of time and resources.

I did a tiny bit of testing for fun (only one run each, so take with a grain of salt):
Two harvesters on one initial gem patch with one gem mine gave 4400 in the first game minute.
Two harvesters on one initial gem patch with one gem mine and available ore patch in the distance gave 4425 in the first minute. (Current case.)
Two harvesters on one initial gem patch with one gem mine and available ore patch in the distance micro-ed to harvest gems first gave 4750 in the first minute. (Proposed case.)

The difference is not much, but it would accumulate over the game time (this is only one minute, and the initial patch had a value of 1700).

Setup was the following (in case someone else wants to waste a bit of their time ^^):
![grafik](https://user-images.githubusercontent.com/7704140/64424164-98506700-d0a8-11e9-857f-e4d413f3b5c4.png)

> A true solution would be that the AI actively tries to distribute harvesters between patches but that is a lot more effort than what the options are currently.

I suppose the theory is simple, we could let the AI assign one harvester to each mine after it noticed the initial amount of ore around the mine is gone.

@netnazgul:

> How do you differentiate between an explicit order to gather long distance rather than something close to the ref from a ref scan result? I assume a player expects harvesters to go to the pointed location instead of them trying to be smart.

I do not understand, this is exactly answered in the opening post? If given an explicit order it will (as before) use that as starting point, not the refinery location.

> Another question is - which ref to use as a scan source in this case?

The refinery it just unloaded at. Newly created harvesters can use their own location (I think they do that already) until they unloaded once.

If you want to go one step further, you could have the harvester look for the nearest available refinery before travelling back as well (although that might create some new issues, just throwing the idea out here).

**Comment by GraionDilach** [comment]
It could be my project running a different spreading rate causing my troubles, but I experienced everything I explained for years in this engine with AI harvesters never moving on from one patch to another, and constantly carrying single-bale runs from said closest patch instead. Mind you that most of my reports are coming from observing 6P FFAs daily on my project, so I am very familiar with how AI harvesting acts here.

**Comment by pchote** [comment]
Related: #15829.

**Comment by abcdefg30** [comment]
I am not denying that running single-bale loads is less efficient. The first test shows this actually. This is just saying that there might be an even better solution.

**Comment by abcdefg30** [comment]
Fwiw, I have already received several complaints about this behaviour ingame.

**Comment by deleted-user-1** [comment]
> Fwiw, I have already received several complaints about this behaviour ingame.

That a player can't be confident that a harvester keeps mining a mine directly next to a refinery is a clear regression from a player standpoint (loss of control). 

@abcdefg30's proposal makes sense to me and I hope the old behaviour can be restored. It should be possible to retain the new behavior for the AI while fixing it for the people that actually play the game.

**Comment by PunkPun** [comment]
> That a player can't be confident that a harvester keeps mining a mine directly next to a refinery is a clear regression from a player standpoint (loss of control).

it is not a clear regression because previously harvesters were still unreliable and still had a mind of their own

I don't think that there should be a distinction between explicit and non-explicit orders because player constantly gives explicit orders to ore truck, it would be better to have a distinction between AI and player controlled harvesters

**Comment by deleted-user-1** [comment]
Present in ~~20201213~~ 20200503.
![img](https://user-images.githubusercontent.com/59416324/103135347-fc40d980-46b7-11eb-9cbd-e6b8d851c1d4.png)

Edit: Above example was not played on the playtest.
